nvbench/cupti_profiler.cuh

```diff
@@ -0,0 +1,129 @@
+/*
+ *  Copyright 2021 NVIDIA Corporation
+ *
+ *  Licensed under the Apache License, Version 2.0 with the LLVM exception
+ *  (the "License"); you may not use this file except in compliance with
+ *  the License.
+ *
+ *  You may obtain a copy of the License at
+ *
+ *      http://llvm.org/foundation/relicensing/LICENSE.txt
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+#pragma once
+
+#include <nvbench/config.cuh>
+#include <nvbench/device_info.cuh>
+
+#include <string>
+#include <vector>
+#include <optional>
+
+namespace nvbench::detail
+{
+
+
+#ifdef NVBENCH_HAS_CUPTI
+/**
+ * Pass required metrics in the constructor and organize your code as follows
+ * to get counters back:
+ *
+ * ```
+ * cupti_profiler cupti(
+ *      nvbench::device_info{0},
+ *      { "l1tex__t_sector_hit_rate.pct" });
+ *
+ * cupti->prepare_to_user_loop();
```

Comment:
This example is very helpful!

Small language nitpick about this method name, `prepare_user_loop` sounds more natural.